pkg/util/limiter/query_limiter_test.go

```diff
@@ -188,11 +285,225 @@ func BenchmarkQueryLimiter_AddSeries(b *testing.B) {
+func BenchmarkQueryLimiter_AddSeries_WithCallerDedup_90pct(b *testing.B) {
+	const (
+		metricName   = "test_metric"
+		uniqueSeries = 100
+		totalSeries  = 1000 // 90% duplicates
+	)
+
+	// Create few unique series
+	uniqueSet := make([]labels.Labels, 0, uniqueSeries)
+	for i := 0; i < uniqueSeries; i++ {
+		uniqueSet = append(uniqueSet, labels.FromMap(map[string]string{
+			model.MetricNameLabel: metricName,
+			"series":              fmt.Sprint(i),
+		}))
+	}
+
+	// Build series with many duplicates
+	series := make([]labels.Labels, 0, totalSeries)
+	for i := 0; i < totalSeries; i++ {
+		series = append(series, uniqueSet[i%uniqueSeries])
+	}
+
+	b.ResetTimer()
+	b.ReportAllocs()
+
+	reg := prometheus.NewPedanticRegistry()
+	limiter := NewQueryLimiter(totalSeries, 0, 0, 0, stats.NewQueryMetrics(reg))
+
+	for b.Loop() {
+		for _, s := range series {
+			_, _ = limiter.AddSeries(s, NoopMemoryTracker{})
+		}
+	}
+}
+
+// BenchmarkQueryLimiter_AddSeries_WithCallerDedup benchmarks caller-side deduplication with 50% duplicates
+func BenchmarkQueryLimiter_AddSeries_WithCallerDedup_50pct(b *testing.B) {
+	const (
+		metricName   = "test_metric"
+		uniqueSeries = 500
+		totalSeries  = 1000 // 50% duplicates
+	)
+
+	// Create unique series
+	uniqueSet := make([]labels.Labels, 0, uniqueSeries)
+	for i := 0; i < uniqueSeries; i++ {
+		uniqueSet = append(uniqueSet, labels.FromMap(map[string]string{
+			model.MetricNameLabel: metricName,
+			"series":              fmt.Sprint(i),
+		}))
+	}
+
+	// Create series array with duplicates
+	series := make([]labels.Labels, 0, totalSeries)
+	for i := 0; i < totalSeries; i++ {
+		series = append(series, uniqueSet[i%uniqueSeries])
+	}
+
+	b.ReportAllocs()
+
+	reg := prometheus.NewPedanticRegistry()
+	limiter := NewQueryLimiter(totalSeries, 0, 0, 0, stats.NewQueryMetrics(reg))
+
+	for b.Loop() {
+		for _, s := range series {
+			_, _ = limiter.AddSeries(s, NoopMemoryTracker{})
+		}
+	}
+}
+
+// assertSameLabels checks if two labels.Labels share the same internal data.
+// This is used to verify that AddSeries returns the same labels object for duplicates.
+// This function is implementation-agnostic and works with:
+// - stringlabels (default): stores labels as a single string with length prefixes
+// - slicelabels (build with -tags slicelabels): stores labels as a slice of Label structs
+// - dedupelabels (build with -tags dedupelabels): uses a shared SymbolTable
+//
+// To test with different implementations:
+//
+//	go test -tags slicelabels ./pkg/util/limiter/
+//	go test -tags dedupelabels ./pkg/util/limiter/
+func assertSameLabels(t *testing.T, a, b labels.Labels) {
+	t.Helper()
+
+	aVal := reflect.ValueOf(a)
+	bVal := reflect.ValueOf(b)
+
+	// Try stringlabels implementation (default)
+	// stringlabels stores data in a "data" field of type string
+	if aData := aVal.FieldByName("data"); aData.IsValid() && aData.Kind() == reflect.String {
```

Comment:
Rather than doing this runtime detection of which implementation is in use, what if we used the `slicelabels` build tag to include a different implementation based on which build tag is enabled? 

This is how we handle the different implementations in other places.